### DISCUSSION TITLE
OKD 4.8 pods fail to display logs and even the exec command fails with unauthorized error

### DISCUSSION THREAD

**Discussion by reddybhavaniprasad** [okd-project/okd#1711] [Q&A]
I am using OKD4.8 cluster for my deployments and it use to work properly until last week. Now I am unable to view the logs or unable to exec into the pod.
Following are the errors I am facing,
```
[kubenode@api-int ansible-start]$ oc whoami
waruser
[kubenode@api-int ansible-start]$ oc logs mongo-mongodb-0 -n database
error: You must be logged in to the server (the server has asked for the client to provide credentials ( pods/log mongo-mongodb-0))
[kubenode@api-int ansible-start]$ oc exec -it mongo-mongodb-0 -n database -- mongo
error: unable to upgrade connection: Unauthorized
```
I have successfully logged into the cluster and the user has admin rights. I have checked the certificates as well, they will expire on 2024 but I am unable to find the actual cause of this issue.

I have referred this link for this issue - https://stackoverflow.com/questions/50016954/error-you-must-be-logged-in-to-the-server-the-server-has-asked-for-the-client

Please help me to resolve this issue and let me know if you need any more details.

**Comment by vrutkovs** [comment]
Does `oc login` again fixes this issue?

**Reply by reddybhavaniprasad** [reply]
no I have tried logout and login multiple times but that does not work. I am still facing the unauthorized error.

**Reply by vrutkovs** [reply]
Check if kubeconfig generated during the installation works

**Reply by reddybhavaniprasad** [reply]
I have tried with the kubeconfig file created during the installation. That is also throwing the same error.
```
[kubenode@api-int ansible-start]$ oc logs mongo-mongodb-0 -n database --kubeconfig=install_dir/auth/kubeconfig
error: You must be logged in to the server (Unauthorized)
[kubenode@api-int ansible-start]$ oc whoami
waruser
```

**Reply by vrutkovs** [reply]
In that case see https://access.redhat.com/solutions/4845381 - if you can ssh on the nodes there'll be valid kubeconfigs to access the cluster

**Reply by reddybhavaniprasad** [reply]
I have tried the two approaches mentioned in the above document but both of them are resulting in the same error as mentioned below,

A new kubeconfig file has been generated from this file .openshift_install_state.json and used below,
```
[kubenode@api-int bhprasad]$ oc get po -n database
Error from server (Forbidden): pods is forbidden: User "system:anonymous" cannot list resource "pods" in API group "" in the namespace "database"
[kubenode@api-int bhprasad]$ oc get po -n database --kubeconfig=kubeconfig
NAME                      READY   STATUS             RESTARTS   AGE
mongo-mongodb-0           1/1     Running            0          2d15h
mongo-mongodb-1           1/1     Terminating        0          22h
mongo-mongodb-arbiter-0   0/1     ImagePullBackOff   0          44h
[kubenode@api-int bhprasad]$ oc logs -f mongo-mongodb-0 -n database
Error from server (Forbidden): pods "mongo-mongodb-0" is forbidden: User "system:anonymous" cannot get resource "pods" in API group "" in the namespace "database"
[kubenode@api-int bhprasad]$ oc logs -f mongo-mongodb-0 -n database --kubeconfig=kubeconfig
error: You must be logged in to the server (the server has asked for the client to provide credentials ( pods/log mongo-mongodb-0))
```

kubeconfigs from first master node,

```
[kubenode@api-int node-kubeconfigs]$ oc get po -n database --kubeconfig=lb-ext.kubeconfig
NAME                      READY   STATUS             RESTARTS   AGE
mongo-mongodb-0           1/1     Running            0          2d16h
mongo-mongodb-1           1/1     Terminating        0          23h
mongo-mongodb-arbiter-0   0/1     ImagePullBackOff   0          44h
[kubenode@api-int node-kubeconfigs]$ oc logs mongo-mongodb-0 -n database --kubeconfig=lb-ext.kubeconfig
error: You must be logged in to the server (the server has asked for the client to provide credentials ( pods/log mongo-mongodb-0))
[kubenode@api-int node-kubeconfigs]$ oc logs mongo-mongodb-0 -n database --kubeconfig=lb-int.kubeconfig
error: You must be logged in to the server (the server has asked for the client to provide credentials ( pods/log mongo-mongodb-0))
[kubenode@api-int node-kubeconfigs]$ oc logs mongo-mongodb-0 -n database --kubeconfig=localhost.kubeconfig
error: You must be logged in to the server (the server has asked for the client to provide credentials ( pods/log mongo-mongodb-0))
[kubenode@api-int node-kubeconfigs]$ oc logs mongo-mongodb-0 -n database --kubeconfig=localhost-recovery.kubeconfig
error: You must be logged in to the server (the server has asked for the client to provide credentials ( pods/log mongo-mongodb-0))
```

Both the approaches are failing with the same error while trying to view logs and exec commands in the pods.

**NOTE:** I am able to view the logs of pods from openshift namespaces. This issue is only arising when trying to view the logs of pods that I have deployed.

**Reply by vrutkovs** [reply]
Wait, any kubeconfig can give you a pod list - but neither works on getting the logs of the pod?

**Reply by reddybhavaniprasad** [reply]
yes.. I am able to get the pod list but unable to view the pod logs or exec any commands in it.
But, I am facing this issue only for the namespaces I have created. I am able to view the logs of the pods in openshift namespaces.

**Reply by vrutkovs** [reply]
Check that nodes are Ready - if not, make sure CSRs for these nodes are approved

**Reply by reddybhavaniprasad** [reply]
There are 3 Master nodes and 17 worker nodes in the cluster. All the 3 Master nodes are in Ready state and only 6 out of 17 worker nodes are in Ready state. Also, there are no CSRs displayed to approve.
```
[kubenode@api-int bhprasad]$ oc get csr -A --kubeconfig=kubeconfig
No resources found
```
My understanding is that the oc commands should still work even with some of the worker nodes in Ready state. Please let me know if it is really required to bring all the worker nodes into Ready state in order to view the pod logs.

**Reply by vrutkovs** [reply]
Check why nodes are notready (I assume you can't fetch logs for the pods on NotReady pods) via `oc describe node`. The nodes may need kubelet restarted to make them generate new CSRs

**Reply by reddybhavaniprasad** [reply]
We are currently experiencing problems with our VMWare hosts due to a shortage of CPU resources.
As a result, the nodes are unresponsive, and there is uncertainty regarding the number of nodes that can be restored. 
Despite having some worker nodes in the "Ready" state, we are noticing a lack of Certificate Signing Requests (CSRs). 
Could you also clarify whether the cluster can still be recovered if certain worker nodes remain in the "NotReady" state?

**Reply by vrutkovs** [reply]
CSRs are generated by kubelet on the node (if necessary, i.e. it expires or a new node attempts to join). 

If vSphere makes VMs starve for CPU they won't respond for log requests. Not much OKD can do about that

**Reply by reddybhavaniprasad** [reply]
ok Thanks @vrutkovs I will try to bring up all the worker nodes into Ready state and would reach out to you in case of any issues.

**Reply by reddybhavaniprasad** [reply]
Thanks @vrutkovs I am able to view the logs and be able to execute the commands on the pods after deleting the NotReady nodes and adding them back again.

But I am facing another issue, some of the worker nodes going into "**NotReady**" state frequently. I have observed from the VSphere dashboard that the nodes are going into NotReady state due to **High CPU consumption**.

There is no much load running on the cluster but I have managed to get the processes that are showing high CPU consumption. Please find the screenshot below for the list of processes.

![image](https://github.com/okd-project/okd/assets/31069365/4a97f4ea-3f16-4a17-a0de-00e07d89b275)

Please help me to resolve this issue by making the worker nodes stable.

**Comment by levindecaro** [comment]
Hi, your ceph-mgr memory usage are abnormal, high kswapd usage indicate process are fighting for available memory, it should be one of the reason kubelet being unresponsive.
